### PR TITLE
KT-45875: Add Sequence.randomOrNull and Sequence.random

### PR DESCRIPTION
**Pull request by talemke**
This resolves [KT-45875](https://youtrack.jetbrains.com/issue/KT-45875), by adding `Sequence.randomOrNull` and `Sequence.random` functions. Both allow for a custom random-source to be provided.

The `randomOrNull` function picks a random element out of the sequence and returns `null` if the sequence is empty. This function also works for sequences with nullable elements, I thought it should be up to the end-user to deal with this, as it introduces a conflicting meaning of `null`.

The `random` function picks a random element and throws a `NoSuchElementExcepetion` in case the sequence is empty.

The implementation works as described by Roman Elizarov in the issue, consuming a lot of randomness, but not requiring the sequence to be completely allocated in memory.

## [Checklist](https://github.com/JetBrains/kotlin/blob/master/docs/contributing.md#checklist)

> Before submitting the pull request, make sure that you can say "YES" to each point in this short checklist:

- [x] You provided the link to the related issue(s) from YouTrack
- [x] You made a reasonable amount of changes related only to the provided issues
- [x] You can explain changes made in the pull request
- [x] You ran the build locally and verified new functionality
- [x] You ran related tests locally and they passed
- [x] You do not have merge commits in the pull request

## Proof of correctness

It might be of interest, whether the algorithm actually provides the same probability for all elements in the sequence, and does not prefer, for example, the first or the last elements.

For this, we need to show that given `n ∈ ℕ` elements, each element has a probability of `1 / n`. We can ignore the case of `n = 0`, as we just return `null` or throw an exception in that case.

### Assumption

We assume we have a good random function `random(n)`, which returns any integer in `[1, n]` with equal probability of `1 / n`, for every `n ∈ ℕ`.

**Note:** In the implementation, we use a slightly different function, that returns an integer in `[0, n[`, however, assuming the numbers are still evenly distributed, this does not affect the proof, as we only need to replace `random(n) = 1` with `random(n) = 0` in the implementation.

### Induction Base

The first element, so `n = 1`, should have a probability of `1 / 1 = 100%`. This can be done, by picking the element if `random(n) = random(1) = 1`, which, by definition of the random function, is always the case.

### Induction Step

Given a sequence of `n` elements, for every new element `n + 1`, we need to prove things:

1. The new element has a probability of `1 / (n+1)` to be picked.
2. All previous elements, had a probability of `1 / n`, and now have a probability of `1 / (n+1)`.

#### Proof of the first

We simply pick the new element, if `random(n + 1) = 1`, and thus with a probability of `1 / (n+1)`.

#### Proof of the latter

We can assume, that every previous element, up to this point, has an even probability of `1 / n`. The probability for these elements to be picked now, is that *they have been picked earlier, **and** the new element has **not** been picked*. Expressed mathematically, the probability is `(1 / n) * (1 - (1 / (n+1)))`. We want all these elements to now have a probability of `1 / (n+1)`.

We now need to prove that:
1. `1 / (n+1) = (1 / n) * (1 - (1 / (n+1)))`
2. `1 = (1 / n) * (1 - (1 / (n+1))) * (n+1)`
3. `1 = (1 / n) * ((n+1) - ((n+1) / (n+1)))`
4. `1 = (1 / n) * (n + 1 - 1)`
5. `1 = (1 / n) * n`
6. `1 = 1`
7. This adds up. ✔️ 

#### Conclusion

All elements now have a probability of `1 / (n+1)`, which is exactly what we want.

This small proof might be absolutely not needed, but I found it rather not entirely obvious, that all elements have an equal probability, so I guess it's better to be safe than sorry.

### CODE REVIEW

**Comment by mcpiroman** [comment]
If there was a small buffer of M = say 4, then randomness consumption could be reduced to N/M+1, greatly reducing CPU time as well.
I.e. draw every M items and, if taken, store this chunk in buffer. At the end draw 1 item from the buffer and return it.

The question is whether such a fixed buffer could be stack allocated at given backend. Alternatively can just use 4 variables.
M can also be a default parameter, but this can make it harder for runtime to use stack.

Also, random can be removed for the first item, not to do random of 100% - afaik this case is not optimized in default random impl.

**Comment by TASSIA710** [comment]
Sorry for the late reply, I was quite busy the past few days!

> If there was a small buffer of M = say 4, then randomness consumption could be reduced to N/M+1, greatly reducing CPU time as well.
> I.e. draw every M items and, if taken, store this chunk in buffer. At the end draw 1 item from the buffer and return it.

I do see your point, but I'm not sure if this is wished by the team. The whole idea of this random-implementation for sequences is to not allocate anything. Yes, a few variables/a small list might not be that big of a deal, but then again, how do we settle how large the chunks should be, etc.

I can add a separate random function, that takes a chunk-size parameter and then works through the sequence in chunks. I think for the pure `Sequence.random` function, we should not allocate anything, as it is rather unintuitive in my opinion, but I'll wait for further input by the maintainers on this.

> Also, random can be removed for the first item, not to do random of 100% - afaik this case is not optimized in default random impl.

Technically yeah, but given a sequence of `n = 1000` elements, one "random generation" more or less, does not matter. Removing this would, in my opinion, just make the code harder to read while not really giving any benefit at all.

File: libraries/stdlib/src/kotlin/collections/Sequences.kt [hunk 1]
```diff
@@ -677,3 +677,35 @@ public fun <T : Any> generateSequence(seed: T?, nextFunction: (T) -> T?): Sequen
...
+ * This function works differently in comparison to, for example, [List.random].
+ * It does not allocate a list of all elements in the sequence, however, it does consume
+ * much more randomness (roughly 1 "randomly generated integer" per element in the sequence).
+ *
+ * @param random the source of randomness to use
+ * @return a randomly picked element
+ * @throws NoSuchElementException if the sequence is empty
+ */
+public fun <T> Sequence<T>.random(random: Random = Random): T {
+    return randomOrNull(random) ?: throw NoSuchElementException("Sequence is empty.")
```
Comment (mcpiroman) [inline libraries/stdlib/src/kotlin/collections/Sequences.kt:710]: Won't this throw on null element from not empty sequence?

**Comment by mcpiroman** [comment]
> I think for the pure `Sequence.random` function, we should not allocate anything

Well, currently it uses 1-element (stack allocated) buffer, so it may be M as well. But a second overload seems as the best option, because the impl would be quite different, and the user can choose M.


> Technically yeah, but given a sequence of n = 1000 elements, one "random generation" more or less, does not matter.

But it will for a lot of small sequences, which may also be common. But this is not a big deal in either way.

**Comment by talemke** [comment]
Will close this PR and wait for more input from the maintainers on the related YouTrack issue, regarding the chunked approach and possible solutions regarding infinite sequences. To my knowledge, this feature still isn't marked as approved and it doesn't look like it will be soon :/

**Pull request closed without merging**